samples/nrf9160/http_application_fota/src/main.c

```diff
@@ -0,0 +1,230 @@
+		if (err != 0) {
+			__ASSERT(false,
+				"flash_get_page_info_by_offs returned error %d\n",
+				err);
+		}
+		flash_write_protection_set(flash_dev, false);
+		if (!is_flash_page_erased[info.index]) {
+			err = flash_erase(flash_dev, address, info.size);
+			if (err != 0) {
+				__ASSERT(false,
+					"flash_erase returned error %d at address %08x\n",
+					err, address);
+			}
+			is_flash_page_erased[info.index] = true;
+		}
+		err = flash_write(flash_dev, address, dfu->fragment,
+							dfu->fragment_size);
+		flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			__ASSERT(false,
+				"Flash write error %d at address %08x\n",
+				err, address);
+		}
+		address += dfu->fragment_size;
+		break;
+	}
+
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE:
+		LOG_INF("DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE");
+		break;
+
+	case DOWNLOAD_CLIENT_EVT_ERROR: {
+		LOG_ERR("Download client error");
+		download_client_disconnect(dfu);
+		__ASSERT(false, "Something went wrong, please restart "
+			"the application\n");
+		break;
+	}
+	default:
+		break;
+	}
+	return 0;
+}
+
+
+/**@brief Turn on LED0 and LED1 if CONFIG_APPLICATION_VERSION
+ * is 2 and LED0 otherwise.
+ */
+static void led_app_version(void)
+{
+	struct device *dev;
+
+	dev = device_get_binding(LED_PORT);
+	__ASSERT(dev != 0, "Nordic nRF GPIO driver was not found!\n");
+
+	gpio_pin_configure(dev, LED0_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED0_GPIO_PIN, 1);
+
+#if CONFIG_APPLICATION_VERSION == 2
+	gpio_pin_configure(dev, LED1_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED1_GPIO_PIN, 1);
+#endif
+}
+
+static void dfu_worker_thread(void *p1, void *p2, void *p3)
+{
+	app_dfu_init();
+	app_dfu_transfer_start();
+
+	while (dfu.status != DOWNLOAD_CLIENT_STATUS_DOWNLOAD_COMPLETE) {
+		download_client_process(&dfu);
+	}
+
+	download_client_disconnect(&dfu);
+	LOG_INF("Download complete");
+}
+
+void dfu_button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		    u32_t pins)
+{
+	LOG_INF("dfu_button_pressed");
+	start_dfu = true;
+}
+
+static void dfu_button_init(void)
+{
+	int err;
+	struct device *gpiob;
+
+	gpiob = device_get_binding(SW0_GPIO_CONTROLLER);
+	__ASSERT(gpiob != 0, "Nordic nRF GPIO driver was not found!\n");
+	err = gpio_pin_configure(gpiob, SW0_GPIO_PIN,
+			GPIO_DIR_IN		|
+			GPIO_INT		|
+			GPIO_PUD_PULL_UP|
+			GPIO_INT_EDGE	|
+			GPIO_INT_ACTIVE_LOW);
+	if (err == 0) {
+		gpio_init_callback(&gpio_cb, dfu_button_pressed,
+			BIT(SW0_GPIO_PIN));
+		err = gpio_add_callback(gpiob, &gpio_cb);
+	}
+	if (err == 0) {
+		err = gpio_pin_enable_callback(gpiob, SW0_GPIO_PIN);
+	}
+	__ASSERT(err == 0, "Unable to configure SW0_GPIO_PIN!\n");
+}
+
+static void start_dfu_thread(void)
+{
+	start_dfu = false;
+	k_thread_create(&dfu_thread, dfu_thread_stack,
+		K_THREAD_STACK_SIZEOF(dfu_thread_stack),
+		dfu_worker_thread,	NULL, NULL, NULL,
+		K_PRIO_COOP(7), 0, K_NO_WAIT);
+}
+
+
+int main(void)
+{
+	boot_write_img_confirmed();
+	start_dfu = false;
```

Comment:
Not really needed but maybe start_dfu could be a [semaphore](https://docs.zephyrproject.org/latest/reference/kernel/synchronization/semaphores.html) or a [mutex](https://docs.zephyrproject.org/latest/reference/kernel/synchronization/mutexes.html) just to showcase more of zephyr's multi-threading capabilites. Then you could do give/take or lock/unlock to enable the start of dfu and you could do the thread initialization(creation) in the start of the program.

As it stands now I think if you press the button multiple times you could create more than 1 DFU thread by getting a `dfu_button_pressed` callback as `start_dfu` is being set to false before the thread is created. However I think `k_thread dfu_thread would limit this but I'm unsure`

Reply:
Thread removed.